Build failure: python311Packages.cupy

### Steps To Reproduce

```
nix-shell -p '(import <nixpkgs> {
          inherit system;
          config.allowUnfree = true;
          config.cudaSupport = true;
          config.cudaCapabilities = [
            "7.5"
            "8.0"
          ];}).python311Packages.cupy'
```

### Build log

full log: https://gist.github.com/cfhammill/22616c79dfe5a1d19755bf0eb51cfddf

seemingly relevant sections include
```
-------- Configuring Module: cusparselt --------
/build/tmp8adp7qy1/a.cpp:1:10: fatal error: cusparseLt.h: No such file or directory
    1 | #include <cusparseLt.h>
      |          ^~~~~~~~~~~~~~
compilation terminated.
command '/nix/store/mpm3i0sbqc9svfch6a17179fs64dz2kv-gcc-wrapper-13.3.0/bin/g++' failed with exit code 1
```

and

```
Exception: Could not find libcudart_static.a: /nix/store/cmy1ismvlzgw5qjxr6an84kysgfbc3yj-cudatoolkit-joined-11.8/lib64/libcudart_static.a does not exist
```

which is interesting because replacing `lib64` with `lib` in the path above does point to a statically compiled libcudart.

### Notify maintainers

@samuela @SomeoneSerge @hyphon81 

### Metadata

Please run `nix-shell -p nix-info --run "nix-info -m"` and paste the result.

```console
[user@system:~]$ nix-shell -p nix-info --run "nix-info -m"
 - system: `"x86_64-linux"`
 - host os: `Linux 5.15.0-101-generic, Ubuntu, 22.04.3 LTS (Jammy Jellyfish), nobuild`
 - multi-user?: `yes`
 - sandbox: `yes`
 - version: `nix-env (Nix) 2.18.1`
 - nixpkgs: `/mnt/data/home/HammillC/.nix-defexpr/channels/nixpkgs`
```

---

Add a :+1: [reaction] to [issues you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[issues you find important]: https://github.com/NixOS/nixpkgs/issues?q=is%3Aissue+is%3Aopen+sort%3Areactions-%2B1-desc

They're pretty much all the same

```
cupy_backends/cuda/libs/cutensor.cpp:7219:79: error: ‘cutensorPlan_t’ was not declared in this scope; did you mean ‘cutensorAlgo_t’?                                                                               7219 |         __pyx_v_status = cutensorReduce(((cutensorHandle_t)__pyx_v_handle), ((cutensorPlan_t)__pyx_v_plan), ((void *)__pyx_v_alpha), ((void *)__pyx_v_A), ((void *)__pyx_v_beta), ((void *)__pyx_v_C), ((void *)__pyx_v_D), ((void *)__pyx_v_workspace), __pyx_v_workspaceSize, ((cudaStream_t)__pyx_v_stream));
      |                                                                               ^~~~~~~~~~~~~~                                                                                                                    |                                                                               cutensorAlgo_t                                                                                                              cupy_backends/cuda/libs/cutensor.cpp:7219:94: error: expected ‘)’ before ‘__pyx_v_plan’                                                                                                                            7219 |         __pyx_v_status = cutensorReduce(((cutensorHandle_t)__pyx_v_handle), ((cutensorPlan_t)__pyx_v_plan), ((void *)__pyx_v_alpha), ((void *)__pyx_v_A), ((void *)__pyx_v_beta), ((void *)__pyx_v_C), ((void *)__pyx_v_D), ((void *)__pyx_v_workspace), __pyx_v_workspaceSize, ((cudaStream_t)__pyx_v_stream));
      |                                                                             ~                ^~~~~~~~~~~~                                                                                                       |                                                                                              )                                                                                                            cupy_backends/cuda/libs/cutensor.cpp: In function ‘PyObject* __pyx_f_13cupy_backends_4cuda_4libs_8cutensor_destroyOperationDescriptor(intptr_t, int)’:                                                            cupy_backends/cuda/libs/cutensor.cpp:7482:63: error: ‘cutensorOperationDescriptor_t’ was not declared in this scope; did you mean ‘cutensorContractionDescriptor_t’?
 7482 |         __pyx_v_status = cutensorDestroyOperationDescriptor(((cutensorOperationDescriptor_t)__pyx_v_desc));                                                                                                     |                                                               ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~                                                                                                                     |                                                               cutensorContractionDescriptor_t                                                                                                             cupy_backends/cuda/libs/cutensor.cpp:7482:93: error: expected ‘)’ before ‘__pyx_v_desc’                                                                                                                            7482 |         __pyx_v_status = cutensorDestroyOperationDescriptor(((cutensorOperationDescriptor_t)__pyx_v_desc));                                                                                                     |                                                             ~                               ^~~~~~~~~~~~                                                                                                        |                                                                                             )            
```

my guess is the cutensor version isn't high enough.

--- a/pkgs/development/python-modules/cupy/default.nix
+++ b/pkgs/development/python-modules/cupy/default.nix
@@ -41,14 +41,14 @@ let
 in
 buildPythonPackage rec {
   pname = "cupy";
-  version = "13.0.0";
+  version = "13.2.0";
   format = "setuptools";
 
   disabled = pythonOlder "3.7";
 
   src = fetchPypi {
     inherit pname version;
-    hash = "sha256-LwTnhX9pKnEzYNycOwZwmAarhAT8o5ta+XIcBKKXmq4=";
+    hash = "sha256-5NvSsu1BWaXMDA+YpxCgFJUOssFu60VelWEo87O9DVE=";
   };
 
   # See https://docs.cupy.dev/en/v10.2.0/reference/environment.html. Seting both

--- a/pkgs/development/python-modules/cupy/default.nix
+++ b/pkgs/development/python-modules/cupy/default.nix
@@ -1,8 +1,8 @@
 {
   lib,
   buildPythonPackage,
-  fetchPypi,
-  cython,
+  fetchFromGitHub,
+  cython_0,
   fastrlock,
   numpy,
   wheel,
@@ -13,6 +13,7 @@
   addOpenGLRunpath,
   pythonOlder,
   symlinkJoin,
+  fetchpatch
 }:
 
 let
@@ -46,10 +47,21 @@ buildPythonPackage rec {
 
   disabled = pythonOlder "3.7";
 
-  src = fetchPypi {
-    inherit pname version;
-    hash = "sha256-5NvSsu1BWaXMDA+YpxCgFJUOssFu60VelWEo87O9DVE=";
+  src = fetchFromGitHub {
+    owner = "cupy";
+    repo = "cupy";
+    rev = "v13.2.0";
+    hash = "sha256-vZAtpIZztmsYeJeuq7yl7kgZse2azrIM3efHDmUswJI=";
+    fetchSubmodules = true;
   };
+
+  patches = [
+    (fetchpatch {
+      url =
+        "https://github.com/cfhammill/cupy/commit/67526c756e4a0a70f0420bf0e7f081b8a35a8ee5.patch";
+      hash = "sha256-WZgexBdM9J0ep5s+9CGZriVq0ZidCRccox+g0iDDywQ=";
+    })
+  ];
 
   # See https://docs.cupy.dev/en/v10.2.0/reference/environment.html. Seting both
   # CUPY_NUM_BUILD_JOBS and CUPY_NUM_NVCC_THREADS to NIX_BUILD_CORES results in
@@ -65,7 +77,7 @@ buildPythonPackage rec {
     setuptools
     wheel
     addOpenGLRunpath
-    cython
+    cython_0
     cudaPackages.cuda_nvcc
   ];
 
@@ -78,7 +90,6 @@ buildPythonPackage rec {
 
   NVCC = "${lib.getExe cudaPackages.cuda_nvcc}"; # FIXME: splicing/buildPackages
   CUDA_PATH = "${cudatoolkit-joined}";
-  LDFLAGS = "-L${cudaPackages.cuda_cudart}/lib/stubs";
 
   propagatedBuildInputs = [
     fastrlock

--- a/pkgs/development/python-modules/cupy/default.nix
+++ b/pkgs/development/python-modules/cupy/default.nix
@@ -18,9 +18,7 @@
 
 let
   inherit (cudaPackages) cudnn cutensor nccl;
-  cudatoolkit-joined = symlinkJoin {
-    name = "cudatoolkit-joined-${cudaPackages.cudaVersion}";
-    paths = with cudaPackages; [
+  outpaths = with cudaPackages; [
       cuda_cccl # <nv/target>
       cuda_cccl.dev
       cuda_cudart
@@ -37,7 +35,10 @@ let
 
       # Missing:
       # cusparselt
-    ];
+  ];
+  cudatoolkit-joined = symlinkJoin {
+    name = "cudatoolkit-joined-${cudaPackages.cudaVersion}";
+    paths = outpaths ++ lib.concatMap (f: lib.map f outpaths) [lib.getLib lib.getDev (lib.getOutput "static") (lib.getOutput "stubs")];
   };
 in
 buildPythonPackage rec {

--- a/pkgs/development/python-modules/cupy/default.nix
+++ b/pkgs/development/python-modules/cupy/default.nix
@@ -43,7 +43,7 @@ let
 in
 buildPythonPackage rec {
   pname = "cupy";
-  version = "13.2.0";
+  version = "13.3.0";
   format = "setuptools";
 
   disabled = pythonOlder "3.7";
@@ -51,8 +51,8 @@ buildPythonPackage rec {
   src = fetchFromGitHub {
     owner = "cupy";
     repo = "cupy";
-    rev = "v13.2.0";
-    hash = "sha256-vZAtpIZztmsYeJeuq7yl7kgZse2azrIM3efHDmUswJI=";
+    rev = "refs/tags/v${version}";
+    hash = "sha256-eQZwOGCaWZ4b0JCHZlrPHVQVXQwSkibHb02j0czAMt8=";
     fetchSubmodules = true;
   };
 

